// dodopayments/resources/discounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
import { DefaultPageNumberPagination } from "../pagination.mjs";
export class Discounts extends APIResource {
    /**
     * If `code` is omitted or empty, a random 16-char uppercase code is generated.
     */
    create(body, options) {
        return this._client.post('/discounts', { body, ...options });
    }
    /**
     * GET /discounts/{discount_id}
     */
    retrieve(discountId, options) {
        return this._client.get(`/discounts/${discountId}`, options);
    }
    /**
     * PATCH /discounts/{discount_id}
     */
    update(discountId, body, options) {
        return this._client.patch(`/discounts/${discountId}`, { body, ...options });
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/discounts', DiscountsDefaultPageNumberPagination, { query, ...options });
    }
    /**
     * DELETE /discounts/{discount_id}
     */
    delete(discountId, options) {
        return this._client.delete(`/discounts/${discountId}`, {
            ...options,
            headers: { Accept: '*/*', ...options?.headers },
        });
    }
}
export class DiscountsDefaultPageNumberPagination extends DefaultPageNumberPagination {
}
Discounts.DiscountsDefaultPageNumberPagination = DiscountsDefaultPageNumberPagination;
