// dodopayments/resources/payments.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
import { DefaultPageNumberPagination } from "../pagination.mjs";
export class Payments extends APIResource {
    create(body, options) {
        return this._client.post('/payments', { body, ...options });
    }
    retrieve(paymentId, options) {
        return this._client.get(`/payments/${paymentId}`, options);
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/payments', PaymentListResponsesDefaultPageNumberPagination, {
            query,
            ...options,
        });
    }
    retrieveLineItems(paymentId, options) {
        return this._client.get(`/payments/${paymentId}/line-items`, options);
    }
}
export class PaymentListResponsesDefaultPageNumberPagination extends DefaultPageNumberPagination {
}
Payments.PaymentListResponsesDefaultPageNumberPagination = PaymentListResponsesDefaultPageNumberPagination;
