// dodopayments/resources/payments.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentListResponsesDefaultPageNumberPagination = exports.Payments = void 0;
const resource_1 = require("../resource.js");
const core_1 = require("../core.js");
const pagination_1 = require("../pagination.js");
class Payments extends resource_1.APIResource {
    create(body, options) {
        return this._client.post('/payments', { body, ...options });
    }
    retrieve(paymentId, options) {
        return this._client.get(`/payments/${paymentId}`, options);
    }
    list(query = {}, options) {
        if ((0, core_1.isRequestOptions)(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/payments', PaymentListResponsesDefaultPageNumberPagination, {
            query,
            ...options,
        });
    }
    retrieveLineItems(paymentId, options) {
        return this._client.get(`/payments/${paymentId}/line-items`, options);
    }
}
exports.Payments = Payments;
class PaymentListResponsesDefaultPageNumberPagination extends pagination_1.DefaultPageNumberPagination {
}
exports.PaymentListResponsesDefaultPageNumberPagination = PaymentListResponsesDefaultPageNumberPagination;
Payments.PaymentListResponsesDefaultPageNumberPagination = PaymentListResponsesDefaultPageNumberPagination;
