// dodopayments/resources/subscriptions.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../resource.mjs";
import { isRequestOptions } from "../core.mjs";
import { DefaultPageNumberPagination } from "../pagination.mjs";
export class Subscriptions extends APIResource {
    create(body, options) {
        return this._client.post('/subscriptions', { body, ...options });
    }
    retrieve(subscriptionId, options) {
        return this._client.get(`/subscriptions/${subscriptionId}`, options);
    }
    update(subscriptionId, body, options) {
        return this._client.patch(`/subscriptions/${subscriptionId}`, { body, ...options });
    }
    list(query = {}, options) {
        if (isRequestOptions(query)) {
            return this.list({}, query);
        }
        return this._client.getAPIList('/subscriptions', SubscriptionListResponsesDefaultPageNumberPagination, {
            query,
            ...options,
        });
    }
    changePlan(subscriptionId, body, options) {
        return this._client.post(`/subscriptions/${subscriptionId}/change-plan`, {
            body,
            ...options,
            headers: { Accept: '*/*', ...options?.headers },
        });
    }
    charge(subscriptionId, body, options) {
        return this._client.post(`/subscriptions/${subscriptionId}/charge`, { body, ...options });
    }
}
export class SubscriptionListResponsesDefaultPageNumberPagination extends DefaultPageNumberPagination {
}
Subscriptions.SubscriptionListResponsesDefaultPageNumberPagination =
    SubscriptionListResponsesDefaultPageNumberPagination;
