// dodopayments/src/resources/license-key-instances.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from "../resource.js";
import { isRequestOptions } from "../core.js";
import * as Core from "../core.js";
import { DefaultPageNumberPagination, type DefaultPageNumberPaginationParams } from "../pagination.js";

export class LicenseKeyInstances extends APIResource {
  /**
   * @example
   * ```ts
   * const licenseKeyInstance =
   *   await client.licenseKeyInstances.retrieve('lki_123');
   * ```
   */
  retrieve(id: string, options?: Core.RequestOptions): Core.APIPromise<LicenseKeyInstance> {
    return this._client.get(`/license_key_instances/${id}`, options);
  }

  /**
   * @example
   * ```ts
   * const licenseKeyInstance =
   *   await client.licenseKeyInstances.update('lki_123', {
   *     name: 'name',
   *   });
   * ```
   */
  update(
    id: string,
    body: LicenseKeyInstanceUpdateParams,
    options?: Core.RequestOptions,
  ): Core.APIPromise<LicenseKeyInstance> {
    return this._client.patch(`/license_key_instances/${id}`, { body, ...options });
  }

  /**
   * @example
   * ```ts
   * // Automatically fetches more pages as needed.
   * for await (const licenseKeyInstance of client.licenseKeyInstances.list()) {
   *   // ...
   * }
   * ```
   */
  list(
    query?: LicenseKeyInstanceListParams,
    options?: Core.RequestOptions,
  ): Core.PagePromise<LicenseKeyInstancesDefaultPageNumberPagination, LicenseKeyInstance>;
  list(
    options?: Core.RequestOptions,
  ): Core.PagePromise<LicenseKeyInstancesDefaultPageNumberPagination, LicenseKeyInstance>;
  list(
    query: LicenseKeyInstanceListParams | Core.RequestOptions = {},
    options?: Core.RequestOptions,
  ): Core.PagePromise<LicenseKeyInstancesDefaultPageNumberPagination, LicenseKeyInstance> {
    if (isRequestOptions(query)) {
      return this.list({}, query);
    }
    return this._client.getAPIList('/license_key_instances', LicenseKeyInstancesDefaultPageNumberPagination, {
      query,
      ...options,
    });
  }
}

export class LicenseKeyInstancesDefaultPageNumberPagination extends DefaultPageNumberPagination<LicenseKeyInstance> {}

export interface LicenseKeyInstance {
  id: string;

  business_id: string;

  created_at: string;

  license_key_id: string;

  name: string;
}

export interface LicenseKeyInstanceUpdateParams {
  name: string;
}

export interface LicenseKeyInstanceListParams extends DefaultPageNumberPaginationParams {
  /**
   * Filter by license key ID
   */
  license_key_id?: string | null;
}

LicenseKeyInstances.LicenseKeyInstancesDefaultPageNumberPagination =
  LicenseKeyInstancesDefaultPageNumberPagination;

export declare namespace LicenseKeyInstances {
  export {
    type LicenseKeyInstance as LicenseKeyInstance,
    LicenseKeyInstancesDefaultPageNumberPagination as LicenseKeyInstancesDefaultPageNumberPagination,
    type LicenseKeyInstanceUpdateParams as LicenseKeyInstanceUpdateParams,
    type LicenseKeyInstanceListParams as LicenseKeyInstanceListParams,
  };
}
